### ISSUE TITLE
Serious slowness when passing void ptr* in Struct as Buffer

### ISSUE THREAD

**Issue by ggryschuk** [Koromix/koffi#162]
As noted in my other issue, I've encountered serious slow performance when using koffi to load a function that populates a 'void*' member of a struct when passing the struct as a Buffer. I"m not sure that koffi is doing anything 'wrong', heck if I know how it actually ends up populating the 'void *' member as a Buffer in this use case but I'll try to describe the issue and perhaps you can advise either a different way of handling it or that I'll just have to 'bite the bullet' and update my code base to conform to the 'koffi way' of passing javascript objects for use as structs.

Assume we have a C-Function with the following signature:

```
struct DcmElement_t {
	unsigned short group;
	unsigned short element;
	unsigned long VM;
	const char* VR;
	unsigned int length;
	void* valueBuf;
	const char* name;
};

bool GetElementFromDataSet(void** dataSet, struct DcmElement_t* element,unsigned long elementNumber);
```

Here the 'void** dataSet' is a 'pointer to pointer to an opaque' object and that seems to load just fine when passing a Buffer instead of a null member array. elementNumber is 'like' an index in to an array and isn't ulitmately relevant but noted for completeness.

Now in javascript I define the 'struct' here twice, once to use with ref-napi/ref-struct-di to use 'Buffers as pointers' and once for defining the signature of the struct for Koffi, which would look as follows:

```
const DcmElement_t = ref_struct({
	group: ref.types.ushort,
	element: ref.types.ushort,
	VM: ref.types.ulong,
	VR: ref.types.CString,
	length: ref.types.uint,
	valueBuf: ref.refType(void),      <---- NOTE:  This is the important element causing the speed issue
	name: ref.types.CString,
})

const KDcmElement_t = koffi.struct({
	group: koffi.types.ushort,
	element: koffi.types.ushort,
	VM: koffi.types.ulong,
	VR: koffi.types.string,
	length: koffi.types.uint,
	valueBuf: koffi.pointer(koffi.types.void),              <--- again the 'offending' element
	name: koffi.types.string,
})

const KDcmElement_tPtr = koffi.pointer(KDcmElement_t);
```

The 'void** dataSet' opaque handle is defined as:

`cont KDataSetType_PtrPtr = koffi.pointer(koffi.pointer(koffi.opaque));
`

and I tell Koffi how to load the GetElementFromDataSet function as:

`const GetElementFromDataSet = lib.func('GetElementFromDataSet','bool',[KDataSetType_PtrPtr,koffi.inout(KDcmElement_tPtr),ulong]);
`

Now, assuming I've already populated the dataSet as a pointer-to-pointer to opaque by passing a Buffer rather then a [null](the function that does this runs just fine and seems no slower then using ffi-lxe/ffi-napi).

With that all set up if I do:

```
let element = new DcmElement_t
let elementPtr = element.ref();
GetElementFromDataSet(dataSet,elementPtr,1);

```
And I console.log 'element.valueBuf' from this I get:

> element.valueBuf: <Buffer@0x7ff3ab15d020 type: { size: 0, indirection: 1, get: <ref *1> { [Function: get] [length]: 2, [name]: 'get', [prototype]: { [constructor]: [Circular *1] } }, set: <ref *2> { [Function: set] [length]: 3, [name]: 'set', [prototype]: { [constructor]: [Circular *2] } }, name: 'void' }>

This runs SIGNIFICANTLY slower (10x times slower) then if I do:

```
let element = {}
GetElementFromDataSet(dataSet,element,1);

```
and if I console log 'element.valueBuf' from this I get:

> element. valueBuf: [External: 7ff3ab15d020],

So while the 'koffi way' usage of passing a pure javascript object for the 'struct DcmElement*' parameter of the function call works WAY faster I then had to figure out how to convert that 'External Data' object returned by Koffi to a Buffer as the rest of my code that decodes the data 'pointed to' by 'valueBuf' assumes it is a Buffer.

This took a bit of testing and it may not be 'kosher' in that I'm manipulating the storage of a memory address and clearly here only works for a Little Endian machine but the following 'worked' to convert the memory pointed to as an 'External Data' object to a Buffer (or 'ref Buffer' acting as a 'pointer to void')

```
let addr = Number(koffi.address(element.valueBuf));
let valueBufPtr = ref.alloc(ref.refType('void'));
ref.writeInt64LE(valueBufPtr,0,addr);
element.valueBuf =  valueBufPtr.deref();

```
This adds additional processing time to the usage of the function using the 'koffi way' of passing a javascript object as a 'pointer to struct' but depending on where I put this in my code it does still show some speed up vs passing the 'ref-struct' Buffer directly but its still roughly 4x slower then the ffi-lxe/ffi-napi version. This wouldn't be so bad other then this code runs ALOT in what I'm doing & results in a significant performance difference to the main code.

My apologies if this isn't very clear and I realize I'm assuming you are somewhat familiar with the ref-napi/ref-struct-di semantics of using Buffers as 'pointers to data types' so if not then all good & I'll see what I can do about converting my main code base to treating 'structs' as pure javascript objects as I see no hope in moving forward long term with ffi-napi which seems to have been abandoned.

I do know I can use 'koffi.decode' on valueBuf when its returned as [External <number>] in the javascript object (I minimally tested that to see if I understood what [External <number>] was) to retrieve the data as a javscript 'value' and if I have to convert my code to doing that I will but thought I'd post this here in hopes you might have some advise or maybe you can pull off some 'magic' again & update koffi to run faster when populating embedded 'void *' type members of a Buffer passed as a 'pointer to struct'.

O, 1 rather 'simple' question. Do you know of any way in javascript to test that a variable is 'External Data' so I could do something like

```
if(valueBuf instanceof '[External Data]') {

<do something here vs if valueBuf is a Buffer>

}

```
Thanks.

**Comment by Koromix** [comment]
While you mix and match the node-ffi/ref stuff with Koffi, it will be slow because these [modules are slow](https://koffi.dev/benchmarks). In fact, this is the main reason I started to work on Koffi two years ago.

In Koffi, pointer values are turned into External objects, except for `char *` (which Koffi decodes as UTF-8 to JS strings).

In your case, I think you could directly get valueBuf as a Uint8Array by defining the struct like this:

```js
const KDcmElement_t = koffi.struct({
	group: koffi.types.ushort,
	element: koffi.types.ushort,
	VM: koffi.types.ulong,
	VR: koffi.types.string,
	length: koffi.types.uint,
	valueBuf: koffi.array(koffi.types.uint8_t, koffi.sizeof(koffi.types.uintptr_t)),
	name: koffi.types.string,
})
```

Hopefully I got this right.

**Comment by ggryschuk** [comment]
Thanks for the follow up. I'll give that a shot or at least I might try it in other areas. For this specific purpose I already implemented a 'work around' which is really me just implementing the 'koffi way' where I updated my 'decoding routines' to expect an '[External Data]' value, actually tested as '!(valueBuf instanceof Buffer)', since I don't know how to test that a Javascript variable is an 'External Data' object (any thoughts on that?).

My 'work around' did show a significant speed increase where at least the  performance is 'reasonable'.

It was just 1 update in a larger code base so overall using koffi is still somewhat slower (though as noted 'reasonable' or 'usable') so I have other updates I can probably make & using your 'trick' above in defining a void* may prove useful in other areas.

thanks.
